homeassistant/components/amcrest/__init__.py

```diff
@@ -160,47 +205,94 @@ def setup(hass, config):
         else:
             authentication = None
 
-        hass.data[DATA_AMCREST][name] = AmcrestDevice(
-            camera, name, authentication, ffmpeg_arguments, stream_source,
+        hass.data[DATA_AMCREST]['devices'][name] = AmcrestDevice(
+            api, authentication, ffmpeg_arguments, stream_source,
             resolution)
 
         discovery.load_platform(
-            hass, 'camera', DOMAIN, {
+            hass, CAMERA, DOMAIN, {
                 CONF_NAME: name,
             }, config)
 
         if binary_sensors:
             discovery.load_platform(
-                hass, 'binary_sensor', DOMAIN, {
+                hass, BINARY_SENSOR, DOMAIN, {
                     CONF_NAME: name,
                     CONF_BINARY_SENSORS: binary_sensors
                 }, config)
 
         if sensors:
             discovery.load_platform(
-                hass, 'sensor', DOMAIN, {
+                hass, SENSOR, DOMAIN, {
                     CONF_NAME: name,
                     CONF_SENSORS: sensors,
                 }, config)
 
         if switches:
             discovery.load_platform(
-                hass, 'switch', DOMAIN, {
+                hass, SWITCH, DOMAIN, {
                     CONF_NAME: name,
                     CONF_SWITCHES: switches
                 }, config)
 
-    return len(hass.data[DATA_AMCREST]) >= 1
+    if not hass.data[DATA_AMCREST]['devices']:
+        return False
+
+    async def async_extract_from_service(service):
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids == ENTITY_MATCH_ALL:
+            return [entity for entity in hass.data[DATA_AMCREST]['cameras']
```

Comment:
We should use our dispatch helper to signal the correct entity instead of storing the entity in a shared container. Entities are a platform concern.

Reply:
Sorry, but I'm not sure I'm following you. This is an amcrest service, not a camera service. And it doesn't apply to all cameras, just amcrest cameras. How could I use the dispatcher helper to do figure out all the amcrest cameras that exist?

Reply:
In `async_added_to_hass` connect a armcrest unique signal per service to the corresponding entity method.

Reply:
Yes, that makes sense. I've seen the dispatcher methods before. I just didn't think they applied here. Actually I didn't even think of using them because, again, the component was already using `hass.data` before I came on the scene. 😄 I can work on changing this as you suggest.

Reply:
I started looking into making this change, looking at existing example uses of the dispatcher mechanism, and it became clear that I totally misunderstood what you were asking me to do.

It seems you're asking me to no longer keep a list of amcrest camera entities at all. And, hence, you don't want the "hub" service handler to call the camera's (or cameras') service method directly. But rather the camera should connect itself to one signal per service (so a total of ten per camera.) And then the hub should just use the dispatcher to send the service's corresponding signal (with call args if necessary) to "wake up" the camera(s) to run the corresponding service method. Am I even close to understanding now?

If so, I'm not seeing how that will work.

How does the hub "signal the correct entity"? Doesn't it just send a signal, and wouldn't that cause all the cameras that are "listening" on that signal to respond and call the service method? How would they know whether or not they were the "correct" entity, or one of the "correct" entities? Where does the "async_extract_from_service" processing happen? It can't happen in the hub anymore, because it doesn't have the list of amcrest cameras anymore. And how can it happen in the camera if it doesn't have access to the call information? Or do I need to pass the call information along with the dispatcher signal and other possible args? And, if so, does that mean all the camera entities have to individually run the equivalent "async_extract_from_service" processing? Seems inefficient.

Can you point me to a specific, existing use of the dispatcher that is equivalent to what you want me to do here? I found a lot of examples but I didn't seem to find any like this.

Reply:
@MartinHjelmare, after giving it a lot more thought, and looking at more examples, I think I now understand what I need to do. I'll be pushing two more commits. The first will include the rebase and all the other requested changes, and the second will implement the use of the dispatcher. Hopefully it's what you're looking for! :smile: